### PR TITLE
[Development] Add options to restart form & wizard

### PR DESCRIPTION
**Pull request by Mottie**
## Description

When the user restarts a form, the form is redirected to the first page of the form. This change adds two options to the `formConfig.saveInProgress` object ensuring that the wizard completion flag is cleared and the form restarts on the desired page - most like the form introduction page

This PR also updates the `ApplicationStatus` component to accept a `wizardStatus` property, so that it can properly set the wizard as completed when the user resumes, and clears the complete state when the user restarts.

### Intro page render
If the wizard is rendered on the introduction page, use this example:

```js
const formConfig = {
  // ...
  // Defaults to WIZARD_STATUS in 'applications/static-pages/wizard'. Added
  // here if your form uses a unique sessionStorage key
  wizardStorageKey: 'wizardStatus',
  saveInProgress: {
    // *** Form restart-specific options ***
    // Return the destination route ('/' is the introduction page) or null to
    // continue with the default behavior
    restartFormCallback: () => '/',
  },
  // ...
};
```

### App render

If the wizard is rendered in the form app, separate from the `RoutedSavableApp`, then use this code:

```js
const formConfig = {
  // ...
  // Defaults to WIZARD_STATUS in 'applications/static-pages/wizard'. Added
  // here if your form uses a unique sessionStorage key
  wizardStorageKey: 'wizardStatus',
  // ...
};
```

Inside the App itself, you'll need to use the `restartShouldRedirect` helper function:

```js
import WizardContainer from './containers/WizardContainer';
import { WIZARD_STATUS } from './constants'; // should be unique to the form
import {
  WIZARD_STATUS_NOT_STARTED,
  WIZARD_STATUS_COMPLETE,
  WIZARD_STATUS_RESTARTED,
  restartShouldRedirect,
} from 'applications/static-pages/wizard';

export const FormEntry = ({
  const [wizardState, setWizardState] = useState(
    sessionStorage.getItem(WIZARD_STATUS) || WIZARD_STATUS_NOT_STARTED
  );

  const hasRestarted = () => {
    setWizardStatus(WIZARD_STATUS_RESTARTED);
    sessionStorage.setItem(WIZARD_STATUS, value);
    router.push('/'); // Intro page URL
  };

  // ...
  useEffect(() => {
    if (restartShouldRedirect(WIZARD_STATUS)) {
      hasRestarted();
    }
    // ...
  });

  if (wizardState !== WIZARD_STATUS_COMPLETE) {
    return <WizardContainer />;
  }
  // ...
});
```

Related:
- Ticket: https://github.com/department-of-veterans-affairs/va.gov-team/issues/13814
- Doc update: https://github.com/department-of-veterans-affairs/veteran-facing-services-tools/pull/586

## Testing done

Updated unit tests

## Acceptance criteria
- [x] Add ability to go to an alternative page and reset the wizard when the form is restarted
- [x] Add functionality to restart form 526 and 996 (HLR)
- [ ] Verify this change will work with other forms that need support (to do)

## Definition of done
- [ ] Events are logged appropriately
- [x] Documentation has been updated, if applicable
- [x] A link has been provided to the originating GitHub issue (or connected to it via ZenHub)
- [x] No sensitive information (i.e. PII/credentials/internal URLs/etc.) is captured in logging, hardcoded, or specs

### CODE REVIEW

**Comment by Mottie** [comment]
List of forms that need to be tested:
- [x] 526/BDD
- [x] HLR
- [ ] Education VA Form 1990
- [ ] Education VA Form 1990N
- [ ] Education VA Form 1990E
- [ ] Education VA Form 1995
- [ ] Education VA Form 5490
- [ ] Education VA Form 5495
- [ ] Education VA Form 10203
- [ ] Education VA Form 0994
- [ ] VRE Chapt 31 (VA Form 28-1900)
- [ ] VRE Chapt 36 (VA Form 28-8832)
- [ ] FSR VA Form 5655

File: src/applications/disability-benefits/all-claims/config/form.js [hunk 1]
```diff
@@ -151,6 +156,8 @@ const formConfig = {
         'Your saved disability compensation application (21-526EZ) has expired. If you want to apply for disability compensation, please start a new application.',
       saved: 'Your disability compensation application has been saved.',
     },
+    // optional, redirect is in Form5256EZApp
+    // restartFormCallback: () => '/',
```
Comment (timwright12) [inline src/applications/disability-benefits/all-claims/config/form.js:160]: intentional?

**Review by timwright12 — APPROVED** [review]
LGTM, nice commenting!

File: src/applications/disability-benefits/all-claims/config/form.js (hunk 1 above)
Comment (Mottie) [reply src/applications/disability-benefits/all-claims/config/form.js:160]: It is intentionally commented out... it was left in there as sort of documentation, but now I've created a document PR. I'd remove it now, but I really don't want to wait 30+ minutes to rebuild this PR just to remove a comment :D

**Review by ncksllvn — APPROVED** [review]

**Review by ndsprinkle — APPROVED** [review]

**Review by ndsprinkle — APPROVED** [review]

**Review by micahchiang — APPROVED** [review]
l.g.t.m.

File: src/platform/forms/save-in-progress/FormStartControls.jsx [hunk 2]
```diff
@@ -9,6 +9,10 @@ import {
   START_NEW_APP_DEFAULT_MESSAGE,
   APP_TYPE_DEFAULT,
 } from '../../forms-system/src/js/constants';
+import {
+  WIZARD_STATUS,
+  WIZARD_STATUS_RESTARTING,
+} from 'applications/static-pages/wizard';
```
Comment (cvalarida) [inline src/platform/forms/save-in-progress/FormStartControls.jsx:15]: Having the platform code import this module has the downstream effect of importing the `applications/disability-benefits/all-claims/config/selectors` module as well, which then makes platform code—and all applications using this—dependent on code from a single application.

We need to, as much is possible, not have platform code depend on application code.

My recommendation: Pull these constants out into their own module and import them from here and `applications/static-pages/wizard/index.js`.

File: src/platform/forms/save-in-progress/RoutedSavableApp.jsx [hunk 3]
```diff
@@ -20,6 +20,8 @@ import { getSaveInProgressState } from './selectors';
 import environment from 'platform/utilities/environment';
 import { APP_TYPE_DEFAULT } from '../../forms-system/src/js/constants';
 
+import { restartShouldRedirect } from 'applications/static-pages/wizard';
```
Comment (cvalarida) [inline src/platform/forms/save-in-progress/RoutedSavableApp.jsx:23]: Same story here. To eliminate the application dependency from platform code, put this function in another module which _doesn't_ import application code, and then import it from there.

File: src/platform/forms/save-in-progress/RoutedSavableApp.jsx [hunk 4]
```diff
@@ -112,7 +114,20 @@ class RoutedSavableApp extends React.Component {
...
+      let newRoute;
+      const { formConfig = {} } = newProps;
+      const { saveInProgress = {} } = formConfig;
+      if (
+        newProps.isStartingOver &&
+        typeof saveInProgress.restartFormCallback === 'function' &&
+        restartShouldRedirect(formConfig.wizardStorageKey || 'formStatus')
+      ) {
+        // Restart callback can optionally return a new route
+        newRoute = saveInProgress?.restartFormCallback() || null;
```
Comment (cvalarida) [inline src/platform/forms/save-in-progress/RoutedSavableApp.jsx:126]: Blech. Optionally return a new route? I'd prefer to see a consistent return type to reduce our Surface Area for Weirdness. (A phrase which I may hereafter officially coin! :sweat_smile:)

What are the use cases for returning a new route vs not returning a new route?

Comment (Mottie) [reply src/platform/forms/save-in-progress/RoutedSavableApp.jsx:126]: Originally when I made these changes, the `restartFormCallback` was called more than once. Now it should only get called once, so yeah, making it always return a route is a good idea.

File: src/platform/forms/save-in-progress/FormStartControls.jsx (hunk 2 above)
Comment (Mottie) [reply src/platform/forms/save-in-progress/FormStartControls.jsx]: Added a new folder + files in `platform/site-wide/wizard`

File: src/platform/forms/save-in-progress/RoutedSavableApp.jsx (hunk 3 above)
Comment (Mottie) [reply src/platform/forms/save-in-progress/RoutedSavableApp.jsx]: Now in `platform/site-wide/wizard`

File: src/platform/forms/save-in-progress/RoutedSavableApp.jsx (hunk 4 above)
Comment (Mottie) [reply src/platform/forms/save-in-progress/RoutedSavableApp.jsx]: I updated the comment and removed the `|| null`, but the actual `router.push` still has a fallback to the first page of the form. Should it not do that?

Comment (Mottie) [reply src/platform/forms/save-in-progress/RoutedSavableApp.jsx]: Also, if you want... I've included the docs for this change here: https://github.com/department-of-veterans-affairs/veteran-facing-services-tools/pull/586

**Review by timwright12 — APPROVED** [review]

**Review by timwright12 — APPROVED** [review]

**Review by cvalarida — APPROVED** [review]
Forms library code looks okay. :+1:

**Pull request merged**